src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs

```diff
@@ -270,14 +270,63 @@ internal MethodInfo GetKeyValuePairMethodInfo
             { return _helper.GetKeyValuePairMethodInfo; }
         }
 
+        private ConstructorInfo _nonAttributedTypeConstructor;
+
         /// <SecurityNote>
         /// Critical - fetches information about which constructor should be used to initialize non-attributed types that are valid for serialization
         /// Safe - only needs to be protected for write
         /// </SecurityNote>
         [SecuritySafeCritical]
         internal ConstructorInfo GetNonAttributedTypeConstructor()
         {
-            return _helper.GetNonAttributedTypeConstructor();
+            if (_nonAttributedTypeConstructor == null)
+            {
+                // Cache the ConstructorInfo to improve performance.
+                _nonAttributedTypeConstructor = _helper.GetNonAttributedTypeConstructor();
+            }
+
+            return _nonAttributedTypeConstructor;
+        }
+
+        private Func<object> _makeNewInstance;
+        private Func<object> MakeNewInstance
+        {
+            get
+            {
+                if (_makeNewInstance == null)
+                {
+                    try
+                    {
+                        _makeNewInstance = FastInvokerBuilder.GetMakeNewInstanceFunc(UnderlyingType);
+                    }
+                    catch
```

Comment:
ArgumentException can be thrown from `FastInvokerBuilder.GetMakeNewInstanceFunc` when `UnderlyingType` has no public default ctor. Therefore I think the best way of fixing this actually is to check whether the default ctor is public before calling into this method. I've made changes here and in `CreateNewInstanceViaDefaultConstructor()`. Thanks for catching this.